card/mod/01_history/lib/card/action.rb

```diff
@@ -61,47 +69,46 @@ def delete_old
     # writing here (disabled history), we still have to generate change stream
     # events in another way.
 
-    def changed_fields obj, changed_fields
-      changed_fields.each do |f|
-        Card::Change.create field: f, value: obj[f], card_action_id: id
-      end
+    # def changed_fields obj, changed_fields
+    #   changed_fields.each do |f|
+    #     Card::Change.create field: f, value: obj[f], card_action_id: id
+    #   end
+    # end
+
+    def value field
+      return unless (change = change field)
+      interpret_value field, change.value
     end
 
-    def edit_info
-      @edit_info ||= {
-        action_type:  "#{action_type}d",
-        new_content:  new_values[:content],
-        new_name:     new_values[:name],
-        new_cardtype: new_values[:cardtype],
-        old_content:  old_values[:content],
-        old_name:     old_values[:name],
-        old_cardtype: old_values[:cardtype]
-      }
+    def change field
+      changes[interpret_field field]
     end
 
-    def new_values
-      @new_values ||=
-        {
-          content:  new_value_for(:db_content),
-          name:     new_value_for(:name),
-          cardtype: ((typecard = Card[new_value_for(:type_id).to_i]) &&
-                     typecard.name.capitalize)
-        }
+    def changes
+      @changes ||=
+        begin
+          hash = {}
+          card_changes.each do |change|
+            hash[change.field.to_sym] = change
+          end
+          hash
+        end
     end
```

Comment:
the core of the new approach is that the caching is more on-demand, and further behind the API